src/service/storage-impl.js

```diff
@@ -0,0 +1,288 @@
+   * @return {!Promise}
+   * @override
+   */
+  set(name, value) {
+    return this.saveStore_(store => {
+      store[name] = value;
+    });
+  }
+
+  /**
+   * Removes the specified property. Returns the promise that's resolved when
+   * the operation completes.
+   * @param {string} name
+   * @return {!Promise}
+   * @override
+   */
+  remove(name) {
+    return this.saveStore_(store => {
+      delete store[name];
+    });
+  }
+
+  /**
+   * @return {!Promise<!JSONObject>}
+   * @private
+   */
+  getStore_() {
+    if (this.store_) {
+      return Promise.resolve(this.store_);
+    }
+    return this.whenStarted_.then(() => {
+      return this.binding_.loadStore(this.origin_);
+    }).then(store => {
+      return this.store_ = store;
+    });
+  }
+
+  /**
+   * @param {function(!JSONObject)} mutator
+   * @return {!Promise}
+   * @private
+   */
+  saveStore_(mutator) {
+    return this.getStore_().then(store => {
+      mutator(store);
+      return this.binding_.saveStore(this.origin_, this.store_);
+    }).then(this.broadcastReset_.bind(this));
+  }
+
+  /** @private */
+  listenToBroadcasts_() {
+    this.viewer_.onBroadcast(message => {
+      if (message['type'] == 'amp-storage-reset' &&
+              message['origin'] == this.origin_) {
+        log.fine(TAG, 'Received reset message');
+        this.store_ = null;
+      }
+    });
+  }
+
+  /** @private */
+  broadcastReset_() {
+    log.fine(TAG, 'Broadcasted reset message');
+    this.viewer_.broadcast({
+      'type': 'amp-storage-reset',
+      'origin': this.origin_
+    });
+  }
+}
+
+
+/**
+ * A binding provides the specific implementation of storage technology.
+ * @interface
+ */
+class StorageBindingDef {
+
+  /**
+   * Returns the promise that yields the store blob for the specified origin.
+   * @param {string} unusedOrigin
+   * @return {!Promise<!JSONObject>}
+   */
+  loadStore(unusedOrigin) {}
+
+  /**
+   * Saves the store blob for the specified origin and returns the promise
+   * that's resolved when the operation completes.
+   * @param {string} unusedOrigin
+   * @param {!JSONObject} unusedStore
+   * @return {!Promise}
+   */
+  saveStore(unusedOrigin, unusedStore) {}
+}
+
+
+/**
+ * Storage implementation using Web LocalStorage API.
+ * @implements {StorageBindingDef}
+ * @private Visible for testing only.
+ */
+export class LocalStorageBinding {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const {!Window} */
+    this.win = win;
+  }
+
+  /**
+   * @param {string} origin
+   * @return {string}
+   * @private
+   */
+  getKey_(origin) {
+    return `amp-store:${origin}`;
+  }
+
+  /** @override */
+  loadStore(origin) {
+    return new Promise(resolve => {
+      resolve(this.win.localStorage.getItem(this.getKey_(origin)) || {});
+    });
+  }
+
+  /** @override */
+  saveStore(origin, store) {
+    return new Promise(resolve => {
+      this.win.localStorage.setItem(this.getKey_(origin), store);
+      resolve();
```

Comment:
This gives me automatic rejection if `setItem` method fails without needing to use try/catch directly.